### PR TITLE
Add CSS Transition end output on mat-progress-bar value animation

### PR DESCRIPTION
**Pull request by glouischandra**
Demo of the issue: https://stackblitz.com/edit/matprogressanimation?file=app%2Fprogress-bar-determinate-example.html

### CODE REVIEW

**Comment by glouischandra** [comment]
crisbeto - we need a notifier for when the animation completed to avoid it being cutoff if we're just relying on the value. This demo shows the problem: https://stackblitz.com/edit/matprogressanimation?file=app%2Fprogress-bar-determinate-example.html

If we attach transitionend on mat-progress-bar tag would it be hard to distinguish on which element animation ended? I haven't tried this but would all transitionend on value or bufferValue bubble up together? The event added here is specific for value transition and it feels cleaner that way. Thoughts?

**Comment by crisbeto** [comment]
You can look at the `event.target` to filter out only the events that you care about.

**Comment by glouischandra** [comment]
Thanks Cris, cmiiw: that works but I would need to know the specific details of elements inside mat-progress-bar which I think breaks encapsulation of this component.

File: src/lib/progress-bar/progress-bar.ts [hunk 1]
```diff
@@ -82,6 +88,18 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
   set bufferValue(v: number) { this._bufferValue = clamp(v || 0); }
   private _bufferValue: number = 0;
 
+  @ViewChild("primaryValueBar") primaryValueBar: ElementRef;
+
+  /** 
+   * Event that indicates animation end on value bar. Currently only support determinate and buffer. 
+   * Also note, if multiple event update is thrown in quick succession the last event completion is 
+   * the one firing the output.
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: I would rephrase the comment to say that the event will never be emitted for animations that don't "end" (i.e. 'query' and 'indeterminate'). Also that the event will never fire if animations are disabled.

File: src/lib/progress-bar/progress-bar.ts [hunk 2]
```diff
@@ -113,6 +131,21 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
       return {transform: `scaleX(${scale})`};
     }
   }
+
+  /** Attaches value bar element animation end listener. */
+  ngAfterViewInit() {
+    if(this.mode === 'determinate' || this.mode === 'buffer') {
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: Space after `if` (here and below)

File: src/lib/progress-bar/progress-bar.ts [hunk 3]
```diff
@@ -113,6 +131,21 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
       return {transform: `scaleX(${scale})`};
     }
   }
+
+  /** Attaches value bar element animation end listener. */
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: I would remove this comment since it doesn't always describe what `ngAfterViewInit` does. Really the best thing would be to add another method like `_addAnimationEndListener` and call it here, but I don't feel that strongly

File: src/lib/progress-bar/progress-bar.ts [hunk 4]
```diff
@@ -113,6 +131,21 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
...
+  ngAfterViewInit() {
+    if(this.mode === 'determinate' || this.mode === 'buffer') {
+      this._valueAnimationSubscription = fromEvent(this.primaryValueBar.nativeElement, 'transitionend')
+                                          .subscribe(() => this.valueAnimationEnd.next(this.value));
+    }
+  }
+
+  /** Removes value bar element animation end listener. */
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: This comment can be omitted; it's pretty clear/obvious what `ngOnDestroy` is doing

File: src/lib/progress-bar/progress-bar.spec.ts [hunk 5]
```diff
@@ -118,6 +120,53 @@ describe('MatProgressBar', () => {
       expect(progressElement.componentInstance.mode).toBe('buffer');
     });
   });
+
+  describe('animation trigger on determinate setting', () => {
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.spec.ts]: It would be good to also have a unit test with `NoopAnimationsModule`

File: src/lib/progress-bar/progress-bar.ts (hunk 2 above)
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: This should also check the `_animationMode`

Comment (josephperrott) [inline src/lib/progress-bar/progress-bar.ts]: Rather than only observing for animation events when it is `determinate` or `buffer`, we should always set up the subscription and only emit out `valueAnimationEnd` event when appropriate.  Otherwise, the `mode` of the `MatProgressBar` cannot be changed dynamically and still utilize this event

File: src/lib/progress-bar/progress-bar.ts [hunk 6]
```diff
@@ -113,6 +131,21 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
       return {transform: `scaleX(${scale})`};
     }
   }
+
+  /** Attaches value bar element animation end listener. */
+  ngAfterViewInit() {
+    if(this.mode === 'determinate' || this.mode === 'buffer') {
+      this._valueAnimationSubscription = fromEvent(this.primaryValueBar.nativeElement, 'transitionend')
```
Comment (crisbeto) [inline src/lib/progress-bar/progress-bar.ts]: This needs a couple of things:

1. Since the `transitionend` event bubbles, it means that it could fire if a transition completes for an element inside the track. We should filter out based on whether `event.target === this.primaryValueBar.nativeElement`.
2. The event should be bound outside the `NgZone` and it should re-enter the zone only when it emits.

File: src/lib/progress-bar/progress-bar.ts [hunk 7]
```diff
@@ -82,6 +88,18 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
...
+
+  /** 
+   * Event that indicates animation end on value bar. Currently only support determinate and buffer. 
+   * Also note, if multiple event update is thrown in quick succession the last event completion is 
+   * the one firing the output.
+   */
+  @Output() valueAnimationEnd = new EventEmitter<number>();
+
+  /** Reference to animation end subscription to be unsubscribed on destroy. */
+  private _valueAnimationSubscription: Subscription | null;
```
Comment (crisbeto) [inline src/lib/progress-bar/progress-bar.ts]: You can initialize this to `Subscription.EMPTY`. That way you won't have to null check it everywhere.

File: src/lib/progress-bar/progress-bar.spec.ts [hunk 8]
```diff
@@ -1,8 +1,10 @@
-import {TestBed, async, ComponentFixture} from '@angular/core/testing';
-import {Component} from '@angular/core';
+import {TestBed, async, ComponentFixture, fakeAsync, tick} from '@angular/core/testing';
+import {Component, DebugElement} from '@angular/core';
 import {By} from '@angular/platform-browser';
+import {dispatchFakeEvent} from '@angular/cdk/testing';
 import {Location} from '@angular/common';
 import {MatProgressBarModule} from './index';
+import { MatProgressBar } from './progress-bar';
```
[the remark this answers is not in the document]
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.spec.ts]: done.

File: src/lib/progress-bar/progress-bar.ts [hunk 9]
```diff
@@ -82,6 +88,18 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
   set bufferValue(v: number) { this._bufferValue = clamp(v || 0); }
   private _bufferValue: number = 0;
 
+  @ViewChild("primaryValueBar") primaryValueBar: ElementRef;
```
[the remark this answers is not in the document]
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done.

File: src/lib/progress-bar/progress-bar.ts [hunk 10]
```diff
@@ -82,6 +88,18 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
...
+
+  /** 
+   * Event that indicates animation end on value bar. Currently only support determinate and buffer. 
+   * Also note, if multiple event update is thrown in quick succession the last event completion is 
+   * the one firing the output.
+   */
+  @Output() valueAnimationEnd = new EventEmitter<number>();
```
[the remark this answers is not in the document]
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done.

[the remark this answers is not in the document]
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: Created event interface, looks more flexible now, thanks!

File: src/lib/progress-bar/progress-bar.ts (hunk 1 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done.

File: src/lib/progress-bar/progress-bar.ts (hunk 2 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done.

Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: Oh that's nice tip! Definitely more flexible that way, I guess one possible downside is that the output is always executed on unsupported method, but that seems like small overhead?

File: src/lib/progress-bar/progress-bar.ts (hunk 6 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: Couple questions on these before I make the change:
1. The event is bound to specifically div mat-progress-bar-primary element inside the template, so adding the filter target would be redundant? What am I missing here?
2. Is this more of best practice to run CSS event on NgZone? Would re-enter zone looks like this? https://github.com/angular/material2/blob/c2fc3f45b5ccdbff3ade56e0685b621688e9cb6d/src/cdk/text-field/autofill.ts#L72.

File: src/lib/progress-bar/progress-bar.ts (hunk 4 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done.

File: src/lib/progress-bar/progress-bar.ts (hunk 2 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: Ok added, to the binding to only bind transition on _animationMode !== noop.

File: src/lib/progress-bar/progress-bar.spec.ts (hunk 5 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.spec.ts]: I'm not completely following the test where NoopAnimationsModule here, would it look like:
1. Check that event is not bound?
2. On animation === noop, always call output on value change? 

Or would it be something else?

File: src/lib/progress-bar/progress-bar.ts (hunk 6 above)
Comment (crisbeto) [reply src/lib/progress-bar/progress-bar.ts]: 1. That's true, but it's more of a future-proofing thing in case we decide to put something into the element.
2. It's only really necessary if we end up doing point 1, otherwise it would be redundant.

File: src/lib/progress-bar/progress-bar.spec.ts (hunk 5 above)
Comment (jelbourn) [reply src/lib/progress-bar/progress-bar.spec.ts]: I want to ensure that we have a test that covers the code path where `NoopAnimationsModule` is loaded and someone tries to register this event. The goal being to ensure that no error is thrown more so than that a particular outcome occurs.

File: src/lib/progress-bar/progress-bar.ts [hunk 11]
```diff
@@ -82,6 +94,18 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
   set bufferValue(v: number) { this._bufferValue = clamp(v || 0); }
   private _bufferValue: number = 0;
 
+  @ViewChild('primaryValueBar') primaryValueBar: ElementRef;
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: This property should be prefixed with an underscore to indicate that it is not part of the public API surface

Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done

File: src/lib/progress-bar/progress-bar.ts [hunk 12]
```diff
@@ -82,6 +94,18 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
...
+
+  /**
+   * Event that indicates animation end on value bar. This event will not be emitted for
+   * animations that doesn't end, i.e. indeterminate and query. Also if animation is
+   * disabled, this event will not emit.
```
[the remark this answers is not in the document]
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done

File: src/lib/progress-bar/progress-bar.ts [hunk 13]
```diff
@@ -113,6 +137,25 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
...
+  }
+
+  ngOnDestroy() {
+    this._animationEndSubscription.unsubscribe();
+  }
+
+  /** Callback function on animation end, emit animationEnd event on determinate and buffer mode. */
+  private emitAnimationEnd(): void {
+    if (this.mode === 'determinate' || this.mode === 'buffer') {
+      this.animationEnd.next({ value: this.value});
```
[the remark this answers is not in the document]
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done

File: src/lib/progress-bar/progress-bar.ts (hunk 6 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: Awesome, thanks for the explanation. I've updated the latest commit with these two points incorporated. PTAL.

File: src/lib/progress-bar/progress-bar.spec.ts (hunk 5 above)
Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.spec.ts]: Thanks that makes sense. I added some more test with NoopAnimations module turned on. Also I make the animationEnd fire when setting the value in NoopAnimations mode, my thoughts is that page author might be relying on this output to update their DOM in their test so we should fire the output synchronously when there's no animation. PTAL.

File: src/lib/progress-bar/progress-bar.ts [hunk 14]
```diff
@@ -27,6 +35,11 @@ export class MatProgressBarBase {
   constructor(public _elementRef: ElementRef) { }
 }
 
+/** Last animation end data. */
+export interface AnimationEndData {
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.ts]: Ah, one last thing I missed earlier- this should be `ProgressAnimationEnd`; having `Data` at the end is inconsistent with other events around the library

Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: Done.

File: src/lib/progress-bar/progress-bar.ts [hunk 15]
```diff
@@ -113,6 +150,31 @@ export class MatProgressBar extends _MatProgressBarMixinBase implements CanColor
...
+            .subscribe(_ => this._ngZone.run(() => this.emitAnimationEnd()));
+      }));
+    }
+  }
+
+  ngOnDestroy() {
+    this._animationEndSubscription.unsubscribe();
+  }
+
+  /** Callback function on animation end, emit animationEnd event on determinate and buffer mode. */
```
Comment (josephperrott) [inline src/lib/progress-bar/progress-bar.ts]: Method comment should not describe when its used, only what it does.  Should instead be:
```
/** Emit an animationEnd event if in determinate or buffer mode. */
```

Comment (glouischandra) [reply src/lib/progress-bar/progress-bar.ts]: done.
